// @oclif/plugin-update/lib/commands/update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const inquirer_1 = require("inquirer");
const path = (0, tslib_1.__importStar)(require("path"));
const semver_1 = require("semver");
const update_1 = require("../update");
class UpdateCommand extends core_1.Command {
    async run() {
        const { args, flags } = await this.parse(UpdateCommand);
        const updater = new update_1.Updater(this.config);
        if (flags.available) {
            const index = await updater.fetchVersionIndex();
            const allVersions = (0, semver_1.sort)(Object.keys(index)).reverse();
            const localVersions = await updater.findLocalVersions();
            const table = allVersions.map(version => {
                const location = localVersions.find(l => path.basename(l).startsWith(version)) || index[version];
                return { version, location };
            });
            core_1.CliUx.ux.table(table, { version: {}, location: {} });
            return;
        }
        if (args.channel && flags.version) {
            this.error('You cannot specifiy both a version and a channel.');
        }
        return updater.runUpdate({
            channel: args.channel,
            autoUpdate: flags.autoupdate,
            force: flags.force,
            version: flags.interactive ? await this.promptForVersion(updater) : flags.version,
        });
    }
    async promptForVersion(updater) {
        const choices = (0, semver_1.sort)(Object.keys(await updater.fetchVersionIndex())).reverse();
        const { version } = await (0, inquirer_1.prompt)({
            name: 'version',
            message: 'Select a version to update to',
            type: 'list',
            choices: [...choices, new inquirer_1.Separator()],
        });
        return version;
    }
}
exports.default = UpdateCommand;
UpdateCommand.description = 'update the <%= config.bin %> CLI';
UpdateCommand.args = [{ name: 'channel', optional: true }];
UpdateCommand.examples = [
    {
        description: 'Update to the stable channel:',
        command: '<%= config.bin %> <%= command.id %> stable',
    },
    {
        description: 'Update to a specific version:',
        command: '<%= config.bin %> <%= command.id %> --version 1.0.0',
    },
    {
        description: 'Interactively select version:',
        command: '<%= config.bin %> <%= command.id %> --interactive',
    },
    {
        description: 'See available versions:',
        command: '<%= config.bin %> <%= command.id %> --available',
    },
];
UpdateCommand.flags = {
    autoupdate: core_1.Flags.boolean({ hidden: true }),
    available: core_1.Flags.boolean({
        char: 'a',
        description: 'Install a specific version.',
    }),
    version: core_1.Flags.string({
        char: 'v',
        description: 'Install a specific version.',
        exclusive: ['interactive'],
    }),
    interactive: core_1.Flags.boolean({
        char: 'i',
        description: 'Interactively select version to install. This is ignored if a channel is provided.',
        exclusive: ['version'],
    }),
    force: core_1.Flags.boolean({
        description: 'Force a re-download of the requested version.',
    }),
};
