// @salesforce/templates/lib/utils/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TemplateType = void 0;
/**
 * Available Template types
 * Each template type must have a corresponding generator class:
 * - generator class file should locate in generators/
 * - generator class file should default export a generator class extending SfdxGenerator
 * - generator class file should have a name same as the type name, except with the first letter lowercased
 */
var TemplateType;
(function (TemplateType) {
    TemplateType[TemplateType["AnalyticsTemplate"] = 0] = "AnalyticsTemplate";
    TemplateType[TemplateType["ApexClass"] = 1] = "ApexClass";
    TemplateType[TemplateType["ApexTrigger"] = 2] = "ApexTrigger";
    TemplateType[TemplateType["LightningApp"] = 3] = "LightningApp";
    TemplateType[TemplateType["LightningComponent"] = 4] = "LightningComponent";
    TemplateType[TemplateType["LightningEvent"] = 5] = "LightningEvent";
    TemplateType[TemplateType["LightningInterface"] = 6] = "LightningInterface";
    TemplateType[TemplateType["LightningTest"] = 7] = "LightningTest";
    TemplateType[TemplateType["Project"] = 8] = "Project";
    TemplateType[TemplateType["VisualforceComponent"] = 9] = "VisualforceComponent";
    TemplateType[TemplateType["VisualforcePage"] = 10] = "VisualforcePage";
    TemplateType[TemplateType["StaticResource"] = 11] = "StaticResource";
})(TemplateType = exports.TemplateType || (exports.TemplateType = {}));
