// @oclif/plugin-update/lib/hooks/init.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.init = void 0;
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const cross_spawn_1 = (0, tslib_1.__importDefault)(require("cross-spawn"));
const fs = (0, tslib_1.__importStar)(require("fs-extra"));
const path = (0, tslib_1.__importStar)(require("path"));
const util_1 = require("../util");
// eslint-disable-next-line unicorn/prefer-module
const debug = require('debug')('cli:updater');
function timestamp(msg) {
    return `[${new Date().toISOString()}] ${msg}`;
}
async function mtime(f) {
    const { mtime } = await fs.stat(f);
    return mtime;
}
const init = async function (opts) {
    if (opts.id === 'update')
        return;
    if (opts.config.scopedEnvVarTrue('DISABLE_AUTOUPDATE'))
        return;
    const binPath = this.config.binPath || this.config.bin;
    const lastrunfile = path.join(this.config.cacheDir, 'lastrun');
    const autoupdatefile = path.join(this.config.cacheDir, 'autoupdate');
    const autoupdatelogfile = path.join(this.config.cacheDir, 'autoupdate.log');
    const clientRoot = this.config.scopedEnvVar('OCLIF_CLIENT_HOME') || path.join(this.config.dataDir, 'client');
    const autoupdateEnv = Object.assign(Object.assign({}, process.env), { [this.config.scopedEnvVarKey('TIMESTAMPS')]: '1', [this.config.scopedEnvVarKey('SKIP_ANALYTICS')]: '1' });
    async function autoupdateNeeded() {
        try {
            const m = await mtime(autoupdatefile);
            let days = 1;
            if (opts.config.channel === 'stable')
                days = 14;
            m.setHours(m.getHours() + (days * 24));
            return m < new Date();
        }
        catch (error) {
            if (error.code !== 'ENOENT')
                core_1.CliUx.ux.error(error.stack);
            if (global.testing)
                return false;
            debug('autoupdate ENOENT');
            return true;
        }
    }
    await (0, util_1.touch)(lastrunfile);
    const clientDir = path.join(clientRoot, this.config.version);
    if (await fs.pathExists(clientDir))
        await (0, util_1.touch)(clientDir);
    if (!await autoupdateNeeded())
        return;
    debug('autoupdate running');
    await fs.outputFile(autoupdatefile, '');
    debug(`spawning autoupdate on ${binPath}`);
    const fd = await fs.open(autoupdatelogfile, 'a');
    fs.write(fd, timestamp(`starting \`${binPath} update --autoupdate\` from ${process.argv.slice(1, 3).join(' ')}\n`));
    (0, cross_spawn_1.default)(binPath, ['update', '--autoupdate'], {
        detached: !this.config.windows,
        stdio: ['ignore', fd, fd],
        env: autoupdateEnv,
    })
        .on('error', (e) => process.emitWarning(e))
        .unref();
};
exports.init = init;
